### PR TITLE
[RLlib] created check_specs decorator, RLModule PR 1/N

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?
submitting [this PR ](https://github.com/ray-project/ray/pull/29449) in pieces:
check_specs decorator can be added to any module method to enforce input/output struct types. This is useful for imposing a certain input/output behavior in RLModule without taking away the flexibility of implementation details from the user. User would also be efficiently informed about what needs to be implemented. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/models/specs/specs_base.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def validate(self, data: Any) -> None:
 
 
 @DeveloperAPI
-class TensorSpecs(SpecsAbstract):
```
Comment (kouroshHakha) [inline rllib/models/specs/specs_base.py:31]: Realized they can represent only one spec, hence renaming :)

File: rllib/utils/nested_dict.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ class NestedDict(Generic[T], MutableMapping[str, Union[T, "NestedDict"]]):
             >>>                             #            'b': {'c': 200, 'd': 300}}
             >>> # Getting elements, possibly nested:
             >>> print(foo_dict['b', 'c'])   # 200
-            >>> print(foo_dict['b']) # IndexError("Use get for partial indexing.")
+            >>> print(foo_dict['b']) # IndexError
```
Comment (sven1977) [inline rllib/utils/nested_dict.py:88]: Dumb question: Is there a reason why we make this seemingly arbitrary distinction between accessing by `get` (no error) vs direct (error)? Why should both not return the sub-dict? How would the user know this difference?

File: rllib/utils/nested_dict.py [hunk 3]
```diff
@@ -181,7 +181,12 @@ def get(
     def __getitem__(self, k: SeqStrType) -> T:
         output = self.get(k)
         if isinstance(output, NestedDict):
-            raise IndexError("Use get for partial indexing.")
+            raise IndexError(
+                f"Key `{k}` is not a complete key in the given "
+                f"{self.__class__.__name__}. It results in a container "
+                f"with subkeys {set(output.keys())}. To get partial indexing, "
```
Comment (sven1977) [inline rllib/utils/nested_dict.py:187]: -> "To use partial indexing and thus retrieve a sub-structure ..."

File: rllib/models/specs/tests/test_check_specs.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        correct_module.check_input_and_output(input_dict)
+
+        # should raise an error (read the implementation of
+        # check_input_and_output_wo_filter)
+        self.assertRaises(
+            ValueError,
+            lambda: correct_module.check_input_and_output_wo_filter(input_dict),
+        )
+
+    def test_cache(self):
```
Comment (sven1977) [inline rllib/models/specs/tests/test_check_specs.py:186]: Very nice test, going the extra mile!

File: rllib/models/specs/tests/test_check_specs.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    def test_incorrect_implementation(self):
+        incorrect_module = IncorrectImplementation()
+        # this should raise an error saying that the output does not match the
+        # output_spec
+        self.assertRaises(
+            ValueError, lambda: incorrect_module.check_input_and_output({"input": 2})
+        )
+
+        # this should not raise an error because output is not forced to be checked
+        incorrect_module.check_only_input({"input": 2})
```
Comment (sven1977) [inline rllib/models/specs/tests/test_check_specs.py:159]: Dumb question, why would the decorator itself not already complain when it's being instantiated b/c of the missing output check? In other words, why is it allowed to have an implementation that doesn't check both, in- and output?

File: rllib/models/specs/specs_jax.py [hunk 6]
```diff
@@ -11,20 +11,20 @@
 
 
 @DeveloperAPI
-class JAXSpecs(TensorSpecs):
-    @override(TensorSpecs)
+class JAXTensorSpec(TensorSpec):
```
Comment (sven1977) [inline rllib/models/specs/specs_jax.py:14]: Why call this JAXTensorSpec, but the others only `XYZSpec` (w/o the "tensor", e.g. "TFSpec")?

File: rllib/models/specs/specs_dict.py [hunk 7]
```diff
@@ -127,4 +138,151 @@ def validate(
 
     @override(NestedDict)
     def __repr__(self) -> str:
-        return f"ModelSpecDict({repr(self._data)})"
+        return f"ModelSpec({repr(self._data)})"
+
+
+def check_specs(
```
Comment (sven1977) [inline rllib/models/specs/specs_dict.py:144]: Super nice! This is going to be dope for early-catching user errors.

**Review by sven1977 — COMMENTED** [review]
Awesome PR! Thanks for being very meticulous about designing these new APIs from the ground up. These will comprise ground-breaking advances for RLlib toward more user-friendliness and transparency.

Just a bunch of questions and nits.

File: rllib/utils/nested_dict.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/utils/nested_dict.py:88]: good point @sven1977, In hindsight I don't see any reason for not returning the sub-nested dict if `__getitem__` is used. It actually confused myself at some point during a later pr.

File: rllib/models/specs/tests/test_check_specs.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/models/specs/tests/test_check_specs.py:159]: `incorrect_module`'s `run` implementation does not have the correct output type. Therefore those functions that enforce output type checking should raise an error and those that don't should just ignore the output spec enforcement. This is only determined when the function is actually run and not during the function decoration itself.  Implementation is only visible to the decorator when the function is invoked.

File: rllib/models/specs/specs_jax.py (hunk 6 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_jax.py:14]: changed everything to `XXXTensorSpec` to be more precise, we may have `XXXDistributionSpecs` down the line too.

File: rllib/utils/nested_dict.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/utils/nested_dict.py]: fixed.

File: rllib/utils/nested_dict.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/utils/nested_dict.py]: I removed partial indexing error all together due the valid comment above. Also update the examples in the docstring  to show that index error is not raised anymore.

**Comment by kouroshHakha** [comment]
@sven1977 Please re-review.

File: rllib/models/specs/tests/test_check_specs.py (hunk 5 above)
Comment (sven1977) [reply rllib/models/specs/tests/test_check_specs.py:159]: Ah, yes, that makes sense! Thanks for clarifying.

**Review by sven1977 — APPROVED** [review]
LGTM.

**Comment by sven1977** [comment]
`test_nested_dict` failing
cc: @kouroshHakha

File: rllib/models/specs/specs_dict.py [hunk 8]
```diff
@@ -1,42 +1,45 @@
...
+        >>> spec_dict = ModelSpec({
         ...     "obs": {
-        ...         "arm":      TensorSpecs("b, d_a", d_a=64),
-        ...         "gripper":  TensorSpecs("b, d_g", d_g=12)
+        ...         "arm":      TensorSpec("b, d_a", d_a=64),
+        ...         "gripper":  TensorSpec("b, d_g", d_g=12)
         ...     },
-        ...     "action": TensorSpecs("b, d_a", h=12),
+        ...     "action": TensorSpec("b, d_a", h=12),
         ...     "action_dist": torch.distributions.Categorical
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:43]: I have a random question. do you intend for these specs to be checkpointed and restored?
if so, maybe a registry of distribution is a good idea.

File: rllib/models/specs/specs_dict.py [hunk 9]
```diff
@@ -107,16 +110,24 @@ def validate(
...
         if exact_match:
             data_spec_missing_keys = data_keys_set.difference(self._keys_set)
             if data_spec_missing_keys:
                 raise ValueError(_MISSING_KEYS_FROM_SPEC.format(data_spec_missing_keys))
 
         for spec_name, spec in self.items():
             data_to_validate = data[spec_name]
-            if isinstance(spec, TensorSpecs):
-                spec.validate(data_to_validate)
+            if isinstance(spec, TensorSpec):
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:123]: for safety, this needs an "else" clause that just raises error if spec is some random data?

File: rllib/models/specs/specs_dict.py [hunk 10]
```diff
@@ -127,4 +138,151 @@ def validate(
...
+                is_tensor = isinstance(spec, TensorSpec)
+
+                if is_mapping:
+                    if not isinstance(data, Mapping):
+                        raise ValueError(
+                            f"{tag} must be a Mapping, got {type(data).__name__}"
+                        )
+                    data = NestedDict(data)
+
+                if should_validate():
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:233]: should you check should_validate() first thing, so we don't waste time building NestedDict, etc?

File: rllib/models/specs/specs_dict.py [hunk 11]
```diff
@@ -127,4 +138,151 @@ def validate(
...
+                                f"Input spec validation failed on "
+                                f"{self.__class__.__name__}.{func.__name__}, "
+                                f"expected {spec.__name__}, got "
+                                f"{type(data).__name__}."
+                            )
+                return data
+
+            input_data_ = input_data
+            if input_spec:
+                input_spec_ = getattr(self, input_spec)()
```
Comment (gjoliver) [inline rllib/models/specs/specs_dict.py:257]: can input_spec_ just be a member variable on self?
or maybe we can simply assume that the specs will be provided by RLModule under some hardcoded key names.
so by default, folks can simply use the decorator without specifying any parameters.

**Review by gjoliver — COMMENTED** [review]
looks pretty good. just some minor issues.

File: rllib/models/specs/specs_dict.py (hunk 8 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_dict.py:43]: I see. This is a good point. This should be incorporated into the RLModule PR then. I'll look into this there.

File: rllib/models/specs/specs_dict.py (hunk 9 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_dict.py:123]: fixed.

File: rllib/models/specs/specs_dict.py (hunk 11 above)
Comment (kouroshHakha) [reply rllib/models/specs/specs_dict.py:257]: That will assume some hard-coded names on the base class and is not the intention of this decorator. The decorator let's the user choose their own spec names. You will see the use-case in the [RLModule PR](https://github.com/ray-project/ray/pull/29449).

File: rllib/models/specs/specs_dict.py (hunk 10 above)
Comment (gjoliver) [reply rllib/models/specs/specs_dict.py:238]: 👌 👌

**Review by gjoliver — APPROVED** [review]

**Comment by kouroshHakha** [comment]
@gjoliver Can we merge this? The failed tests are again not related.

**Comment by gjoliver** [comment]
mem leak test failures are not related.

**Pull request merged**